peer-pod-controller/controllers/peerpod_controller.go

```diff
@@ -0,0 +1,148 @@
+/*
+Copyright Confidential Containers Contributors
+
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/confidential-containers/cloud-api-adaptor/cmd/cloud-api-adaptor/cloudmgr"
+	confidentialcontainersorgv1alpha1 "github.com/confidential-containers/cloud-api-adaptor/peer-pod-controller/api/v1alpha1"
+	"github.com/confidential-containers/cloud-api-adaptor/pkg/adaptor/cloud"
+	"github.com/confidential-containers/cloud-api-adaptor/pkg/adaptor/cloud/aws"
+	"github.com/confidential-containers/cloud-api-adaptor/pkg/adaptor/cloud/libvirt"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+// PeerPodReconciler reconciles a PeerPod object
+type PeerPodReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+var provider cloud.Provider
+
+//+kubebuilder:rbac:groups=confidentialcontainers.org,resources=peerpods,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=confidentialcontainers.org,resources=peerpods/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=confidentialcontainers.org,resources=peerpods/finalizers,verbs=update
+
+func (r *PeerPodReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+	pp := confidentialcontainersorgv1alpha1.PeerPod{}
+
+	if err := r.Get(ctx, req.NamespacedName, &pp); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	if pp.ObjectMeta.GetDeletionTimestamp() == nil { // TODO: consider filter events
+		return ctrl.Result{}, nil
+	}
+
+	if err := r.setProvider(ctx, pp.Spec.CloudProvider); err != nil { // TODO: set only once?
+		return ctrl.Result{}, err
+	}
```

Comment:
If we only need to support a single cloud provider in a Kubernetes cluster, we don't need to set a cloud provider per pod here.

We can specify a cloud provider name as an argument to the controller binary, and create a Provider obect in `mainl.go` like [`cmd/cloud-api-adaptor/main.go`](https://github.com/confidential-containers/cloud-api-adaptor/blob/900b330e4502ba219abcd900bdf7c541539e2a1c/cmd/cloud-api-adaptor/main.go). Then we can pass the Provider object to this controller.

Reply:
I used CLOUD_PROVIDER instead of setting as parameter to the binary